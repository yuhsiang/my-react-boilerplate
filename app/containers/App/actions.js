import {
  APP_INIT_LOAD,
  APP_LOGOUT,
} from './constants';

export const appInitLoad = (domain) => ({
  type: APP_INIT_LOAD,
  payload: {
    domain,
  },
});

export const logoutUser = () => ({
  type: APP_LOGOUT,
});
