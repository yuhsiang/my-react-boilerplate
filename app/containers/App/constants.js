/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

// Redux types
export const KEY = 'APP';

export const APP_INIT_LOAD = `${KEY}/APP_INIT_LOAD`;
export const APP_LOGOUT = `${KEY}/APP_LOGOUT`;

export const DEFAULT_LOCALE = 'en';
