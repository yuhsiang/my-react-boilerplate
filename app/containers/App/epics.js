import {
  ofType,
} from 'redux-observable';
import { of, empty } from 'rxjs';
import {
  flatMap,
  map,
} from 'rxjs/operators';

import { push } from 'connected-react-router';

import {
  removeSession,
} from '@app/services/storage/tokenStorage';

import { showErrorMessage, showHTTPErrorMessage } from '@components/ErrorMessage';

import {
  APP_API_STATUS,
  APP_ERROR_CODE,
} from 'utils/request';

import {
  logoutUser,
} from './actions';

import {
  APP_LOGOUT,
} from './constants';

export const HTTPStatusEpic = (action$) => (
  action$.pipe(
    ofType(APP_API_STATUS),
    flatMap((action) => {
      showHTTPErrorMessage(action.payload);
      if (action.payload === 401) {
        return of(logoutUser());
      }
      return empty();
    })
  )
);

export const ErrorCodeEpic = (action$) => (
  action$.pipe(
    ofType(APP_ERROR_CODE),
    flatMap((action) => {
      showErrorMessage(action.payload.code);
      return empty();
    })
  ));

export const logoutEpic = (action$) => (
  action$.pipe(
    ofType(APP_LOGOUT),
    map(() => {
      removeSession();
      return push('/');
    })
  )
);

export default [
  HTTPStatusEpic,
  ErrorCodeEpic,
  logoutEpic,
];
