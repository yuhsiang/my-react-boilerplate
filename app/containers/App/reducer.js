import { fromJS } from 'immutable';

import meta, {
//   updateMetaLoading,
//   updateMetaDone,
//   updateMetaError,
} from 'Provider/global/meta';

import {
  APP_LOGOUT,
} from './constants';


export const getInitialState = () => fromJS({
  loadingMeta: meta,
});

const reducer = (state = getInitialState(), action) => {
  // const {
  //   error,
  // } = action;
  switch (action.type) {
    case APP_LOGOUT:
      return getInitialState();

    default:
      return state;
  }
};

export default reducer;
