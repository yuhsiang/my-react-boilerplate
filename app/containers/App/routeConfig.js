export default {
  entry: '/',
  loginPage: '/login',
  beauty: '/beauty', // or bbs page
};
