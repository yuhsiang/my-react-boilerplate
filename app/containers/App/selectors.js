import { createSelector } from 'reselect';
import {
  LOADED,
  LOADING,
  ERROR,
} from 'Provider/global/meta';

export const selectApp = (state) => state.get('app');
export const selectUser = (state) => {
  const app = selectApp(state);
  if (!app) {
    return null;
  }
  return app.get('user');
};

const makeSelectEnterpriseDomain = () => createSelector(
  selectApp,
  (globalState) => globalState.get('enterpriseDomain')
);

const selectAppState = () => createSelector(
  selectApp,
  (globalState) => globalState
);

const selectUserState = () => createSelector(
  selectApp,
  (globalState) => globalState.get('user')
);

export const makeSelectIsEnterpriseAdmin = () => createSelector(
  selectApp,
  (globalState) => globalState.getIn(['enterprise', 'admin'])
);

export const selectIsLoadingUser = () => createSelector(
  selectApp,
  (globalState) => globalState.getIn(['userMeta', LOADING])
);

export const selectHasLoadedUser = () => createSelector(
  selectApp,
  (globalState) => globalState.getIn(['userMeta', LOADED])
);

export const makeSelectHasLoadUserError = () => createSelector(
  selectApp,
  (globalState) => globalState.getIn(['userMeta', ERROR])
);

const selectGroups = () => createSelector(
  selectApp,
  (globalState) => globalState.get('groups')
);

const makeSelectHasMultipleGroups = () => createSelector(
  selectApp,
  (globalState) => {
    const groups = globalState.get('groups');
    return !groups || groups.size > 1;
  }
);

const selectActiveGroup = () => createSelector(
  selectApp,
  (globalState) => {
    const id = globalState.get('activeGroup');
    const groups = globalState.get('groups');
    if (groups) {
      return groups.find((item) => item.get('id') === id);
    }
    return null;
  }
);

const makeSelectActiveGroupName = () => createSelector(
  selectApp,
  (globalState) => {
    const id = globalState.get('activeGroup');
    const groups = globalState.get('groups');
    if (groups) {
      const foundGroup = groups.find((item) => item.get('id') === id);
      return foundGroup ? foundGroup.get('name') : '';
    }
    return '';
  }
);

const selectIsLoadingGroup = () => createSelector(
  selectApp,
  (globalState) => globalState.get('isLoadingGroup')
);

export const makeSelectRoute = () => createSelector(
  (state) => state.get('route'),
  (route) => route.get('location'),
);

export const makeSelectEnterpriseName = () => createSelector(
  selectApp,
  (appState) => appState.getIn(['enterprise', 'name'])
);

export {
  selectAppState,
  selectUserState,
  selectGroups,
  selectActiveGroup,
  selectIsLoadingGroup,
  makeSelectActiveGroupName,
  makeSelectEnterpriseDomain,
  makeSelectHasMultipleGroups,
};
