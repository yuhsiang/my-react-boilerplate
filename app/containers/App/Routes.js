import React from 'react';

import { Switch, Route } from 'react-router-dom';
import EntryPage from '@containers/Entry';
import LoginPage from '@containers/LoginPage/Loadable';
import MainPage from '@containers/MainPage/Loadable';

import routeConfig from './routeConfig';

/**
 * Your can add your register page here
 *
 * careful route '/' is exact
 */
export default () => (
  <Switch>
    <Route path={routeConfig.loginPage} component={LoginPage} />
    <Route path={routeConfig.beauty} component={MainPage} />
    <Route exact path={routeConfig.entry} component={EntryPage} />
    <Route path={'*'} component={() => <div>Not Found</div>}/>
  </Switch>
);
