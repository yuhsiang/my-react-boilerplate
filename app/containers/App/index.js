

/**
 *
 * App.react.js
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 *
 * NOTE: while this component should technically be a stateless functional
 * component (SFC), hot reloading does not currently support SFCs. If hot
 * reloading is not a necessity for you then you can refactor it and remove
 * the linting exception.
 */

/**
* withRouter
* https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/guides/redux.md#blocked-updates
*/
// React
import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { GlobalStyle } from 'Styled/Settings/global';

import Routes from './Routes';

const App = ({
  location,
}) => (
  <>
    <GlobalStyle />
    <Routes
      location={location}
    />
  </>
);

App.propTypes = {
  location: PropTypes.shape({
    pathname: PropTypes.string,
  }),
};

App.defaultProps = {
  location: {
    pathname: '',
  },
};


export default withRouter(App);
