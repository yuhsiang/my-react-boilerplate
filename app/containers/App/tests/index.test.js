import React from 'react';
import 'jest-dom/extend-expect';
import { cleanup } from 'react-testing-library';
import { ConnectedRouter } from 'connected-react-router/immutable';

import configureTestStore from '@app/store/configureTestStore';
import { renderWithThemeRedux } from '@app/store/configureReduxTestProvider';
import history from '@app/utils/history';
import createReducer from '@app/reducers';
import App from '../index';

const reducers = createReducer([], history);
const store = configureTestStore(reducers);

describe('<App />', () => {
  afterEach(cleanup);
  it('should render app with loading spinner', async () => {
    const { container } = renderWithThemeRedux(
      <ConnectedRouter history={history}>
        <App />
      </ConnectedRouter>,
      store
    );
    expect(container).toMatchSnapshot();
  });
});
