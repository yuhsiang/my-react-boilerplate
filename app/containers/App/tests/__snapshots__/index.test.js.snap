// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render app with loading spinner 1`] = `
<div>
  <div
    class="sc-bwzfXH igQbjz"
  >
    <div
      class="sc-bdVaJa fgTwBU"
    >
      <span
        class="loader-text"
      >
        Loading
      </span>
      <div
        class="loader-circle"
      />
      <div
        class="loader-circle-orbit"
      >
        <div
          class="loader-circle-line"
        />
      </div>
    </div>
  </div>
</div>
`;
